### PR TITLE
Fix mail box creation error handling and mail domain addition form

### LINKED ISSUE #350: [BUG] Impossibilité de créer un nouveau domaine en appuyant sur Entrée (staging/preprod)
**DESCRIPTION**

En staging et en preprod, quand on veut saisir son propre domaine et valider le formulaire, la page se rafraîchit mais sans avoir créé le nouveau domaine. On voit à nouveau le formulaire à remplir, sans message d'erreur.

**EDIT**

Je précise ici le constat qu'on vient de faire avec Séb : en réalité le problème semble venir de la validation **avec la touche Entrée**, et non en cliquant sur le bouton. C'est donc le comportement déclenché par la touche Entrée qu'il faut corriger.

### LINKED ISSUE #351: [BUG] Impossibilité de créer une boîte mail dans un domaine actif (en staging et préprod)
Quand on remplit le formulaire correspondant, et qu'on valide, il ne se passe rien.

### PR DESCRIPTION
**Pull request by daproclaima**
## Purpose

When we want to create a mailbox for a mail domain, it takes for this mail domain to have a valid secret configured. As far as I understand, this is done manually. So, this is a complex use case to cover with automated tests, which could not be tested with the e2e tests.

Concerning the form to add a mail domain, it could not be submitted when pressing the 'enter' key and would instead update the URL with a query and a page refresh.

## Proposal

- mailbox creation:
  - catch error 500 when the mail domain secret is invalid and display an error cause message
  - catch error 400 and cause when the mail domain secret is empty and display an error cause message

(watch video from 00:30 to 1:15)

https://github.com/user-attachments/assets/908d7b3e-f9fe-468f-adfe-b3ea4d8836c5

https://github.com/user-attachments/assets/5102c304-1ef7-41eb-9e55-bc31c236188d

- mail domain addition:
  -  preventing default behavior on form submission at the "Enter" key down event

https://github.com/user-attachments/assets/724731d7-514c-428f-872c-e9ebae9092a8

https://github.com/user-attachments/assets/3b3b6df7-ff65-4560-ad94-9a7dc710679d

Error catching and cause replacement were already done on the app, but not with a centralized method. Also, no error message used to be shown for the error 500 type (not shown in the video).

Fix #350, resolves #351

### CODE REVIEW

File: src/frontend/apps/desk/src/features/mail-domains/components/ModalAddMailDomain.tsx [hunk 1]
```diff
@@ -8,18 +8,85 @@ import {
...
+          }, [] as string[]);
+
+        causes = parseCauses(error.cause);
+      }
+
+      if (error.status === 500 || !error.cause) {
+        causes = [
+          t(
+            'Your request cannot be processed because the server is experiencing an error. If the problem ' +
+              'persists, please contact our support to resolve the issue: suiteterritoriale@anct.gouv.fr.',
```
Comment (sdemagny) [inline src/frontend/apps/desk/src/features/mail-domains/components/ModalAddMailDomain.tsx:71]: C'est mieux de retirer le point à la fin de l'adresse mail je pense

**Review by sdemagny — COMMENTED** [review]
En attendant le sentry, est-ce qu'on pourrait pas avoir quelque soit le code d'erreur (tous les codes d'erreur) le code d'erreur récupéré et le message associé  affiché sur l'interface ?
Dès qu'on aura sentry on aura des log clairs et on cachera tout ça à l'utilisateur. 
Le but c'est de ne plus avoir d'interface/validation de formulaire qui ne fonctionne pas en silence.
Qu'est-ce que vous en dites ? @AntoLC @daproclaima

**Comment by daproclaima** [comment]
@sdemagny, je suis du même avis. C'est ce que j'ai fait ici: [add mail domain error handling](https://github.com/numerique-gouv/people/blob/011732bbde89bc303179e4ee7e83830e8e553bec/src/frontend/apps/desk/src/features/mail-domains/components/__tests__/useAddMailDomainApiError.tsx#L83) et [create mailbox error handling](https://github.com/numerique-gouv/people/blob/011732bbde89bc303179e4ee7e83830e8e553bec/src/frontend/apps/desk/src/features/mail-domains/components/forms/__tests__/useCreateMailboxApiError.test.tsx#L85)

File: src/frontend/yarn.lock
Comment (AntoLC) [inline src/frontend/yarn.lock:1]: You don't seem to add or update dependencies, this file should not have changes.

File: src/frontend/apps/e2e/__tests__/app-desk/mail-domains-add.spec.ts
Comment (AntoLC) [inline src/frontend/apps/e2e/__tests__/app-desk/mail-domains-add.spec.ts:1]: Why the tests are skipped ?

File: src/frontend/apps/desk/src/features/mail-domains/components/useAddMailDomainApiError.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+          : stateCauses,
+      );
+    }
+  }, [methods, t, error]);
+
+  React.useEffect(() => {
+    if (errorCauses?.length && methods) {
+      methods.setFocus('name');
+    }
+  }, [methods, errorCauses]);
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/features/mail-domains/components/useAddMailDomainApiError.tsx:68]: I have the feeling this useEffect is not necessary, could you add the setFocus somewhere in the first useEffect ?

**Review by AntoLC — COMMENTED** [review]
`useCreateMailboxApiError.tsx` and `useAddMailDomainApiError.tsx` are very similar, I think you can merge them to have just one hook, no?

File: src/frontend/apps/e2e/__tests__/app-desk/mail-domains-add.spec.ts
Comment (daproclaima) [reply src/frontend/apps/e2e/__tests__/app-desk/mail-domains-add.spec.ts:1]: That's an error. I showed Laurent Bossavit how to manipulate one test with debug mode and skip the others.

File: src/frontend/apps/desk/src/features/mail-domains/components/ModalCreateMailbox.tsx [hunk 3]
```diff
@@ -88,27 +93,47 @@ export const CreateMailboxForm = ({
...
+        causes: ['Mailbox with this Local_part and Domain already exists.'],
+        handleError: () => {
+          methods.setError('local_part', {
+            type: 'manual',
+            message: t('This email prefix is already used.'),
+          });
+          methods.setFocus('local_part');
+        },
+      },
+      secret: {
```
Comment (daproclaima) [inline src/frontend/apps/desk/src/features/mail-domains/components/ModalCreateMailbox.tsx:109]: I did not write tests for this part as Marie is working on a modification to replace these two errors.

File: src/frontend/apps/desk/src/features/mail-domains/components/__tests__/ModalCreateMailbox.test.tsx [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+const mockCreateMailbox = jest.fn();
+const mockOnSuccess = jest.fn();
+jest.mock('../../api/useCreateMailbox', () => ({
+  useCreateMailbox: jest.fn().mockImplementation(() => {
+    return useMutation<void, APIError, CreateMailboxParams>({
+      mutationFn: mockCreateMailbox,
+      onSuccess: mockOnSuccess,
+    });
+  }),
+}));
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/features/mail-domains/components/__tests__/ModalCreateMailbox.test.tsx:56]: To test api with jest we use `fetchMock`.

You have some examples in this test
https://github.com/numerique-gouv/people/blob/main/src/frontend/apps/desk/src/features/teams/member-management/__tests__/MemberGrid.test.tsx#L79-L82

File: src/frontend/apps/desk/src/api/__tests__/useAPIError.test.tsx [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+  handleAPIServerError,
+  useAPIError,
+} from '../useAPIError';
+
+jest.mock('react-i18next', () => ({
+  useTranslation: () => ({
+    t: (key: string) => key,
+  }),
+}));
+
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/api/__tests__/useAPIError.test.tsx:16]: Same, use the `AppWrapper`.

File: src/frontend/apps/desk/src/api/useAPIError.tsx [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+export const handleAPIServerError = (
+  causes: string[],
+  serverErrorParams: Omit<ServerErrorParams, 'defaultMessage'> & {
+    defaultMessage: string;
+  },
+): string[] => {
+  causes = causes.length
+    ? [serverErrorParams.defaultMessage, ...causes]
+    : [serverErrorParams.defaultMessage];
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/api/useAPIError.tsx:54]:
```suggestion
  causes.unshift(serverErrorParams.defaultMessage);
```

File: src/frontend/apps/desk/src/features/mail-domains/components/ModalCreateMailbox.tsx [hunk 7]
```diff
@@ -88,27 +93,47 @@ export const CreateMailboxForm = ({
...
+        causes: [
+          "Please configure your domain's secret before creating any mailbox.",
+          `Please check secret of the mail domain ${mailDomain.slug}`,
+        ],
```
Comment (daproclaima) [inline src/frontend/apps/desk/src/features/mail-domains/components/ModalCreateMailbox.tsx:113]: I did not write tests for this part as Marie is working on a modification to replace these two errors.

File: src/frontend/apps/desk/src/features/mail-domains/components/__tests__/ModalAddMailDomain.test.tsx [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+const mockAddMailDomain = jest.fn();
+const mockOnSuccess = jest.fn();
+jest.mock('../../api/useAddMailDomain', () => ({
+  useAddMailDomain: jest.fn().mockImplementation(() => {
+    return useMutation<MailDomain, APIError, string>({
+      mutationFn: mockAddMailDomain,
+      onSuccess: mockOnSuccess,
+    });
+  }),
+}));
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/features/mail-domains/components/__tests__/ModalAddMailDomain.test.tsx:48]: Same fetchMock is tailored for this kind of test.

File: CHANGELOG.md [hunk 9]
```diff
@@ -8,6 +8,13 @@ and this project adheres to
 
 ## [Unreleased]
 
+### Fixed
+
+🐛(frontend) fix mai domain addition submission #355
```
Comment (AntoLC) [inline CHANGELOG.md:13]:
```suggestion
🐛(frontend) fix mail domain addition submission #355
```

**Review by AntoLC — COMMENTED** [review]
A bit more work, you will be a jest hero soon.

File: src/frontend/apps/desk/src/features/mail-domains/components/__tests__/ModalCreateMailbox.test.tsx [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+
+    await user.type(inputFirstName, 'John');
+    await user.type(inputLastName, 'Doe');
+    await user.type(inputLocalPart, 'john.doe');
+
+    act(() => {
+      inputEmailAddress.focus();
+    });
+
+    await user.type(inputEmailAddress, 'john.doe@mail.com{enter}');
```
[the remark this answers is not in the document]
Comment (daproclaima) [reply src/frontend/apps/desk/src/features/mail-domains/components/__tests__/ModalCreateMailbox.test.tsx:241]: I will apply this to assert that the form is submitted using the "Enter" key press. 
However, the test on ModalAddMailDomain asserting this same behavior passes, even though the keypress event triggers on the mail domain name field. I expected to get the same result on this test.

**Review by AntoLC — APPROVED** [review]
👍

**Pull request merged**